src/modules/storefront/editor.tsx

```diff
@@ -0,0 +1,227 @@
+    font-family: '${({ fontFamily }: PrevTitleProps) => fontFamily}', sans;
+    color: ${({ color }: PrevTitleProps) => color};
+  }
+`;
+
+type PrevTextProps = {
+  color: string;
+  fontFamily: string;
+};
+
+export const PrevText = styled(Text)`
+  &.ant-typography {
+    font-family: '${({ fontFamily }: PrevTextProps) => fontFamily}', sans;
+    color: ${({ color }: PrevTextProps) => color};
+  }
+`;
+type PrevCardProps = {
+  bgColor: string;
+};
+
+export const PrevCard = styled(Card)`
+  &.ant-card {
+    border-radius: 12px;
+    height: 100%;
+    display: flex;
+    align-items: center;
+    background-color: ${({ bgColor }: PrevCardProps) => bgColor};
+  }
+`;
+
+export const PageCard = styled(Card)`
+  margin: 70px 0 32px 0;
+`;
+
+export const PrevCol = styled(Col)`
+  margin: 0 0 24px 0;
+`;
+
+export interface FieldData {
+  name: string | number | (string | number)[];
+  value?: any;
+  touched?: boolean;
+  validating?: boolean;
+  errors?: string[];
+}
+
+export interface LocalFile {
+  file: File | undefined;
+  url: string | undefined;
+}
+
+/// Half reverse-engineered, mostly to avoid use of any
+export interface AntdFile {
+  uid: string;
+  name: string;
+  type: string;
+  size: number;
+  percent?: number;
+  status?: 'uploading' | 'done' | 'error' | 'removed';
+  response: LocalFile;
+  xhr?: unknown;
+}
+
+export type FileInput = (ArweaveFile & { file?: undefined }) | LocalFile | AntdFile;
+
+export const popFile = (f: FileInput): (ArweaveFile & { file?: undefined }) | LocalFile => {
+  if (has<'response'>('response', f)) {
+    return f.response;
+  } else {
+    return f;
+  }
+};
+
+export const reduceFieldData = (data: FieldData[]) =>
+  reduce<FieldData, Record<string, any>>(
+    (acc, data) => assocPath(data.name instanceof Array ? data.name : [data.name], data.value, acc),
+    {},
+    data
+  );
+
+export interface StorefrontEditorProps {
+  track: GoogleTracker;
+}
+
+export const validateSubdomainUniqueness = (
+  ar: ArweaveScope,
+  allowPubkey?: string
+): ((rule: RuleObject, subdomain: string | null | undefined) => Promise<void>) => {
+  return async (_, subdomain) => {
+    const storefront = await ar.storefront.find('holaplex:metadata:subdomain', subdomain ?? '');
+
+    if (isNil(storefront)) return;
+    if (allowPubkey && storefront.pubkey === allowPubkey) return;
+
+    throw new Error('The subdomain is already in use.  Please pick another.');
+  };
+};
+
+export const validateArweaveFunds = (
+  arweaveWalletAddress: string,
+  ar: ArweaveScope,
+  setShowARModal: (val: boolean) => void
+): ((rule: RuleObject, [file]: [FileInput?]) => Promise<void>) => {
+  return async (_, [file]) => {
+    if (isNil(file) || has('url', file)) return;
+
+    const canAfford =
+      arweaveWalletAddress && (await ar.wallet.canAfford(arweaveWalletAddress, file.size));
+
+    if (canAfford) return;
+
+    setShowARModal(true);
+
+    throw new Error('Not enough AR funds to cover the upload fee.');
+  };
+};
```

Comment:
I'm up for the validate to be unaware of the modal state changes. Can await validateArweaveFunds in the onValidate actions and toggle the setShowARModal within the view. Trying to separate business logic from UI management.

Reply:
This code actually just needs to go, I plan on removing validateArweaveFunds once I get the new business logic for the Upload component working.